Unity/RuntimeTypes.cs

```diff
@@ -12,30 +12,62 @@ public class RuntimeTypeArray : IReadOnlyList<RuntimeTypeInfo>
         [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
         unsafe delegate NativeTypeArray* GetRuntimeTypesDelegate();
 
+        [UnmanagedFunctionPointer(CallingConvention.ThisCall)]
+        unsafe delegate char* CStrDelegate(IntPtr @this);
+
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        unsafe delegate IntPtr ClassIDToRTTI(int classID);
+
         public RuntimeTypeInfo this[int index] => types[index];
 
         public int Count => types.Count;
 
         unsafe public RuntimeTypeArray(UnityVersion version, SymbolResolver resolver)
         {
-
             NativeTypeArray* runtimeTypes;
-
-            if (version >= UnityVersion.Unity2017_3) 
+            if (version >= UnityVersion.Unity5_5)
             {
-                runtimeTypes  = resolver.Resolve<NativeTypeArray>("?ms_runtimeTypes@RTTI@@0URuntimeTypeArray@1@A");
-            } else
-            {
-                runtimeTypes = resolver.Resolve<NativeTypeArray>("?ms_runtimeTypes@RTTI@@2URuntimeTypeArray@1@A");
-            }
+                if (version >= UnityVersion.Unity2017_3)
+                {
+                    runtimeTypes = resolver.Resolve<NativeTypeArray>("?ms_runtimeTypes@RTTI@@0URuntimeTypeArray@1@A");
+                }
+                else
+                {
+                    runtimeTypes = resolver.Resolve<NativeTypeArray>("?ms_runtimeTypes@RTTI@@2URuntimeTypeArray@1@A");
+                }
 
-            types = new List<RuntimeTypeInfo>(runtimeTypes->Count);
-            for (int i = 0; i < runtimeTypes->Count; i++)
-            {
-                var info = (&runtimeTypes->First)[i];
-                types.Add(new RuntimeTypeInfo(info, version));
+                types = new List<RuntimeTypeInfo>(runtimeTypes->Count);
+                for (int i = 0; i < runtimeTypes->Count; i++)
+                {
+                    var info = (&runtimeTypes->First)[i];
+                    types.Add(new RuntimeTypeInfo(info, resolver, version));
+                }
             }
+            else 
+            {
+                ClassIDToRTTI ClassIDToRTTI;
+                if(version >= UnityVersion.Unity5_4)
+                {
+                    ClassIDToRTTI = resolver.ResolveFunction<ClassIDToRTTI>("?ClassIDToRTTI@Object@@SAPEAURTTI@@H@Z");
+                }
+                else if (version >= UnityVersion.Unity5_0)
+                {
+                    ClassIDToRTTI = resolver.ResolveFunction<ClassIDToRTTI>("?ClassIDToRTTI@Object@@SAPEAURTTI@1@H@Z");
+                } else
+                {
+                    ClassIDToRTTI = resolver.ResolveFunction<ClassIDToRTTI>("?ClassIDToRTTI@Object@@SAPAURTTI@1@H@Z");
+                }
 
+                types = new List<RuntimeTypeInfo>();
+                for(int i = 0; i < 2000; i++)

```

Comment:
Out of curiosity, where does the `2000` come from? Is it the total number of runtime types in Unity 4.x?

Reply:
I found an old type dumper here [here](https://gist.github.com/robert-nix/7db0145e809b692b63f2#file-unitystructgen-cpp) which had 1024, but the versions i were testing had values around 1100-1200, so I made it 2000 just to be safe. The max value could probably be found by running it on 5.4.6. I suspect the way it works in the unity code is that there is a enum constant for the max value, I might try and see if its obtainable at runtime.

Reply:
There is a static variable ms_MaxClassID that looks like it would be accessible at runtime. It seems like it might be based on KLargestEditorClassID which is inlined (but is accessible from the unstripped PDBs). 

![image](https://user-images.githubusercontent.com/3821056/94257367-8a4cdd00-ff6e-11ea-8fc1-e530845a2616.png)

Reply:
I think Unity 4.x includes stripped PDBs by default (stripping started from Unity 5.x I think?), so it should be safe to rely on that.

Reply:
Max Class ID is needed from 5.4 and lower

Reply:
In that case I think hardcoding 2000 is fine, as long as it's either documented in a comment or moved out into a `const int` field/local.